### ISSUE TITLE
MLX auto-fit replaces the configured context length and disagrees with the prompt cache budget

### ISSUE THREAD

**Issue by philippdubach** [lmstudio-ai/mlx-engine#366]
## Summary

The MLX engine reports the auto-fit context length to the application and ignores the context length that the user configured. The prompt cache uses a different rule and keeps the larger of the two values. The two components then disagree about the context length of the same loaded model.

## Environment

- LM Studio 0.4.21+2
- Runtime `mlx-llm-mac-arm64-apple-metal-advsimd` 1.11.0
- Vendor package `app-mlx-generate-mac14-arm64@34`
- MacBook Pro M4, 24 GB unified memory, macOS (Darwin 25.5.0)
- Model `lmstudio-community/Qwen3.8-27B-MLX-4bit` (16.08 GB, arch `qwen3_5`)

## Steps to reproduce

1. Set the model loading guardrails to OFF.
2. Load the model with an explicit context length: `lms load qwen3.8-27b-mlx --context-length 24576`.
3. Run `lms ps`.
4. Send a prompt of about 7,000 tokens to `/v1/chat/completions`.

## Result

`lms ps` shows a context of 4096. The context field in the graphical interface also returns to 4096 after each load. The server rejects the 7,000-token prompt with this message: `number of tokens to keep from the initial prompt is greater than the context length`.

## Logs

```
[context_fit][WARNING]: Model context auto-fit calculated 0 tokens; using the 4,096 token minimum
[context_fit][INFO]: Model context auto-fit: family=qwen3_5 max=262,144 fitted=4,096
  working_set=17.76GiB reserve=3.00GiB safe_ceiling=14.76GiB baseline=14.95GiB
  full_kv=65536B/token prompt_inputs=10240B/token attention=98304B/token
  rotating_peak=0.00GiB fixed_ssm=0.14GiB estimated_peak=15.76GiB
[cache_store][INFO]: VLM prompt cache context target: configured=24,576 fitted=4,096 effective=24,576
```

The last line shows the problem. The engine receives the configured value of 24,576. The prompt cache keeps 24,576. The engine still reports 4,096.

## Cause

Two code paths use the fitted value in different ways.

`BatchedModelKit._load_model()` stores only the fitted value:

```python
self._effective_context_length = fit_batched_vlm_context(
    model=self.model,
    prefill_step_size=self.prefill_step_size,
)
```

`get_runtime_load_info()` in `generate.py` reports that value to the application:

```python
context_length = getattr(model_kit, "effective_context_length", None)
```

`VlmPromptCacheStore.ensure_max_kv_size()` applies a different rule:

```python
self._max_kv_size = max(configured_max_kv_size or 0, max_kv_size)
```

`_load_model()` never reads `max_kv_size`. The constructor passes that value only to `VlmPromptCacheStore`. The reported context length and the budgeted context length differ by construction.

## The 4,096 floor hides a failed fit

On this hardware the fit computes zero tokens. The model baseline is 14.95 GiB. The safe ceiling is 14.76 GiB. `available_prompt_bytes` is zero, and `tokens_that_fit` is zero.

A result of zero means that the model does not fit at all. The floor at `MIN_FITTED_CONTEXT_TOKENS` turns this result into 4,096 tokens. The load then reports success. The user sees a model that loads but cannot accept a useful prompt.

## Suggested fix

Make the two paths agree. Two changes would fix this:

1. Fail the load when the fit computes zero tokens. Report a memory error instead of the 4,096 floor.
2. Report the explicit user setting, or reject it with an error. Do not replace it silently.

The second change needs care. On this hardware the memory arithmetic is correct. A 16 GB model on 24 GB of unified memory leaves about 2.8 GiB below the recommended working set. A larger context does not fit. A clear error helps the user more than a silent floor.

## Note on iogpu.wired_limit_mb

`calculate_context_fit()` reads `max_recommended_working_set_size` from Metal. This value does not follow `sysctl iogpu.wired_limit_mb`. A raised wired limit cannot change the fit.

## Related issues

These reports show the same disagreement between components:

- lmstudio-ai/lmstudio-bug-tracker#2212: the logs show `fitted=4,864` and `effective=6,144`
- lmstudio-ai/lmstudio-bug-tracker#2250: the fit overrides the configured value upward
- lmstudio-ai/lmstudio-bug-tracker#2191: the load reports success when memory is insufficient

**Comment by marcellobonini** [comment]
Quick fix is to roll back to version 1.10.0

**Comment by adamoates** [comment]
Corroborating this from a different angle: a large-memory, no-pressure system where the fit doesn't hit the zero-token floor, plus a follow-on effect on the VLM prompt-cache **disk** budget, and data on whether the "roll back to 1.10.0" workaround costs anything.

## Environment

- LM Studio backend reused by [Bionic](https://elementlabs.ai) (Element Labs' agentic app — same `~/.lmstudio` runtimes/engine, separate app config)
- MacBook Pro M3 Max, 128 GB unified memory, macOS Darwin 25.6.0
- Runtime `mlx-llm-mac-arm64-apple-metal-advsimd`: reproduced on both 1.10.1 (`app-mlx-generate-mac14-arm64@33`) and 1.11.0 (`@34`); **absent** on 1.10.0 (`@32`), which has no `context_fit.py` at all
- Model `qwen3.5-122b-a10b` (arch `qwen3_5_moe`, 69.62 GB, `max_position_embeddings` 262144)
- Requested: `lms load qwen3.5-122b-a10b --context-length 200000 --parallel 4`, guardrails `low`

## Result — override happens even with ample headroom

Unlike the OP's case (fit computes 0 and floors to 4096 under memory pressure), on this hardware the fit computes a *larger* number and silently overrides the smaller requested value — there's no pressure, no floor, no warning, just a different number than what was asked for:

```
[context_fit][INFO]: Model context auto-fit: family=qwen3_5 max=262,144 fitted=247,040
  working_set=107.52GiB reserve=5.38GiB safe_ceiling=102.14GiB baseline=64.82GiB
  full_kv=24576B/token prompt_inputs=6144B/token attention=131072B/token
  rotating_peak=0.00GiB fixed_ssm=0.14GiB estimated_peak=102.18GiB
[cache_store][INFO]: VLM prompt cache context target: configured=200,000 fitted=247,040 effective=247,040
```

`lms ps` confirms it: CONTEXT column reads 247040 no matter what `--context-length` was passed (tried 100000, 131072, 16384 — same fitted=247,040 every time). So this isn't only a small-VRAM edge case; `fit_batched_vlm_context()` has no parameter for the caller's requested length at all, on either branch (floor or normal).

## A second-order effect: the disk cache budget scales with the wrong number

`VlmPromptCacheStore`'s disk budget is derived from the same fitted (not configured) value, so the override doesn't just misreport context — it inflates disk usage:

```
[disk_budget][INFO]: VLM prompt cache disk budget: stage=final target_tokens=247,040 desired_gib=143.74 cap_gib=143.74 limiter=max_kv_size
```

≈0.58 MiB/token × the wrong token count. After pinning to 1.10.0 and reloading at the same `-c 200000`, the honored request drops the budget to:

```
[disk_budget][INFO]: VLM prompt cache disk budget: stage=final cap_gib=116.34 limiter=max_kv_size
```

~27 GiB of extra disk cache headroom allocated purely from the override, with no relation to actual prompt sizes seen. On a workload that fills that cache, this is real SSD wear and disk pressure for a number the user never asked for.

## Confirming the workaround: `lms runtime select …@1.10.0`

```
mlx-llm-mac-arm64-apple-metal-advsimd@1.11.0
mlx-llm-mac-arm64-apple-metal-advsimd@1.10.1
mlx-llm-mac-arm64-apple-metal-advsimd@1.10.0    ✓
```

After the pin + unload/reload: `lms ps` shows `CONTEXT 200000` exactly as requested, no `context_fit` log line appears at all (the code path doesn't exist pre-1.10.1), and the disk budget recalculates correctly (143.74 → 116.34 GiB). Generation speed was unaffected (~39 → 40.6 tok/s, within noise).

## Does the 1.10.0 rollback cost the llguidance tool-call grammar guard?

1.10.1 also introduced an llguidance-based logits processor that grammar-guards `<tool_call>` emission for Qwen3.5/Gemma4. Since this issue's only workaround so far is "roll back to 1.10.0," I ran a regression suite against the rolled-back runtime to check whether that's a real tradeoff: single tool call, nested JSON arguments, sequential tool→result→tool, parallel/multiple calls, a deliberately tool-free prompt, and 30 repeated calls at T=0.7/concurrency=4. **6/6 cases passed, 0 malformed `<tool_call>` leaks, 0 invalid JSON, 0 missing required args** across the run. For this model/workload, 1.10.0 is a clean rollback — no observed tool-calling regression.

Happy to attach the full 1.10.0↔1.10.1 `mlx_engine` diff or raw logs if that's useful for triage.
